File: game/tools/economy.py
"""
Ressourcen-Buchhaltung, Stadt-Einkommen, Einheiten-Wachstum.

Reiner Python-Code, wird vom Simulator und vom Adventure-CLI verwendet.
Spiegelt scripts/core/Economy.cs (C#-Mirror kommt parallel).
"""
from __future__ import annotations

import dataclasses

RESOURCES = ("gold", "wood", "ore", "mercury", "sulfur", "crystal", "gems")


@dataclasses.dataclass
class Resources:
    gold: int = 0
    wood: int = 0
    ore: int = 0
    mercury: int = 0
    sulfur: int = 0
    crystal: int = 0
    gems: int = 0

    def add(self, other: "Resources") -> None:
        for r in RESOURCES:
            setattr(self, r, getattr(self, r) + getattr(other, r))

    def can_afford(self, cost: "Resources") -> bool:
        return all(getattr(self, r) >= getattr(cost, r) for r in RESOURCES)

    def subtract(self, cost: "Resources") -> None:
        for r in RESOURCES:
            setattr(self, r, getattr(self, r) - getattr(cost, r))

    def copy(self) -> "Resources":
        return Resources(**{r: getattr(self, r) for r in RESOURCES})

    def as_dict(self) -> dict[str, int]:
        return {r: getattr(self, r) for r in RESOURCES}


@dataclasses.dataclass
class Town:
    owner: int
    faction: str
    x: int
    y: int
    # Gebaeude: Einfache Stufen. 0 = nicht gebaut, N = Stufe N.
    # Jede Stufe +1 erhoeht den Wachstums-Multiplikator bei der naechsten Wochen-Rekrutierung.
    dwelling_levels: dict[int, int] = dataclasses.field(default_factory=dict)
    has_castle: bool = False
    has_capitol: bool = False
    # Verfuegbare Einheiten im Dwelling (angesammelt pro Woche)
    available_units: dict[str, int] = dataclasses.field(default_factory=dict)

    def daily_income(self) -> Resources:
        gold = 500 if self.has_capitol else (1000 if self.has_castle else 250)
        return Resources(gold=gold)


@dataclasses.dataclass
class Mine:
    owner: int | None
    resource: str  # "gold" | "wood" | "ore" | "crystal" | "gems" | "sulfur" | "mercury"
    x: int
    y: int

    def daily_income(self) -> Resources:
        out = Resources()
        if self.owner is None:
            return out
        amount = 1000 if self.resource == "gold" else 1
        setattr(out, self.resource, amount)
        return out


WEEKLY_GROWTH_BY_TIER = {1: 22, 2: 12, 3: 7, 4: 4, 5: 3, 6: 2, 7: 1}


def apply_weekly_growth(town: Town, units_by_faction: dict[str, list]) -> None:
    """Erhoeht available_units-Counts gemaess Tier-Standard-Growth."""
    for unit in units_by_faction.get(town.faction, []):
        base = WEEKLY_GROWTH_BY_TIER.get(unit.tier, 1)
        # Jede gebaute Dwelling-Stufe erhoeht Growth um 50 Prozent
        level = town.dwelling_levels.get(unit.tier, 0)
        mult = 1.0 + 0.5 * level
        amount = int(base * mult) if level > 0 else 0
        town.available_units[unit.id] = town.available_units.get(unit.id, 0) + amount


def tick_day(
    player_treasury: Resources,
    towns: list[Town],
    mines: list[Mine],
    day_index: int,
    units_by_faction: dict[str, list] | None = None,
) -> None:
    """Wendet Tages-Einkommen an. Wenn Wochenanfang (day_index % 7 == 1), auch Growth."""
    for t in towns:
        player_treasury.add(t.daily_income())
    for m in mines:
        player_treasury.add(m.daily_income())
    if day_index % 7 == 1 and units_by_faction is not None:
        for t in towns:
            apply_weekly_growth(t, units_by_faction)

File: game/scripts/core/Economy.cs
using System.Collections.Generic;

namespace KromeichHeroes.Core;

public sealed class Resources
{
    public int Gold { get; set; }
    public int Wood { get; set; }
    public int Ore { get; set; }
    public int Mercury { get; set; }
    public int Sulfur { get; set; }
    public int Crystal { get; set; }
    public int Gems { get; set; }

    public void Add(Resources other)
    {
        Gold += other.Gold;
        Wood += other.Wood;
        Ore += other.Ore;
        Mercury += other.Mercury;
        Sulfur += other.Sulfur;
        Crystal += other.Crystal;
        Gems += other.Gems;
    }

    public bool CanAfford(Resources cost) =>
        Gold >= cost.Gold && Wood >= cost.Wood && Ore >= cost.Ore
        && Mercury >= cost.Mercury && Sulfur >= cost.Sulfur
        && Crystal >= cost.Crystal && Gems >= cost.Gems;

    public void Subtract(Resources cost)
    {
        Gold -= cost.Gold;
        Wood -= cost.Wood;
        Ore -= cost.Ore;
        Mercury -= cost.Mercury;
        Sulfur -= cost.Sulfur;
        Crystal -= cost.Crystal;
        Gems -= cost.Gems;
    }

    public Resources Copy() => new()
    {
        Gold = Gold, Wood = Wood, Ore = Ore,
        Mercury = Mercury, Sulfur = Sulfur, Crystal = Crystal, Gems = Gems,
    };
}

public sealed class Town
{
    public int Owner { get; set; }
    public string Faction { get; init; } = "";
    public int X { get; init; }
    public int Y { get; init; }
    public Dictionary<int, int> DwellingLevels { get; init; } = new();
    public bool HasCastle { get; set; }
    public bool HasCapitol { get; set; }
    public Dictionary<string, int> AvailableUnits { get; } = new();

    public Resources DailyIncome()
    {
        int gold = HasCapitol ? 500 : (HasCastle ? 1000 : 250);
        return new Resources { Gold = gold };
    }
}

public sealed class Mine
{
    public int? Owner { get; set; }
    public string Resource { get; init; } = "gold";
    public int X { get; init; }
    public int Y { get; init; }

    public Resources DailyIncome()
    {
        var r = new Resources();
        if (Owner is null) return r;
        int amount = Resource == "gold" ? 1000 : 1;
        switch (Resource)
        {
            case "gold":    r.Gold = amount; break;
            case "wood":    r.Wood = amount; break;
            case "ore":     r.Ore = amount; break;
            case "mercury": r.Mercury = amount; break;
            case "sulfur":  r.Sulfur = amount; break;
            case "crystal": r.Crystal = amount; break;
            case "gems":    r.Gems = amount; break;
        }
        return r;
    }
}

public static class Economy
{
    public static readonly IReadOnlyDictionary<int, int> WeeklyGrowthByTier =
        new Dictionary<int, int> { {1,22},{2,12},{3,7},{4,4},{5,3},{6,2},{7,1} };

    public static void ApplyWeeklyGrowth(Town town, IReadOnlyDictionary<string, UnitData> unitsByFaction)
    {
        foreach (var (id, unit) in unitsByFaction)
        {
            if (unit.Faction != town.Faction) continue;
            int level = town.DwellingLevels.TryGetValue(unit.Tier, out var lv) ? lv : 0;
            if (level <= 0) continue;
            int baseGrowth = WeeklyGrowthByTier.TryGetValue(unit.Tier, out var g) ? g : 1;
            int amount = (int)(baseGrowth * (1.0 + 0.5 * level));
            town.AvailableUnits[id] = town.AvailableUnits.TryGetValue(id, out var have) ? have + amount : amount;
        }
    }

    public static void TickDay(
        Resources treasury,
        IList<Town> towns,
        IList<Mine> mines,
        int dayIndex,
        IReadOnlyDictionary<string, UnitData>? allUnits = null)
    {
        foreach (var t in towns) treasury.Add(t.DailyIncome());
        foreach (var m in mines) treasury.Add(m.DailyIncome());
        if (dayIndex % 7 == 1 && allUnits is not null)
            foreach (var t in towns) ApplyWeeklyGrowth(t, allUnits);
    }
}
